save-backend/src/main/kotlin/com/saveourtool/save/backend/service/TestAnalysisService.kt

```diff
@@ -100,42 +151,157 @@ class TestAnalysisService(
+        }
+    }
+
+    /**
+     * Updates the statistics with test executions from this [execution].
+     *
+     * @param execution a newly-finished `Execution`.
+     * @see updateStatisticsInternal
+     */
+    fun updateStatistics(execution: Execution) {
+        val testRunCount: Long
+
+        val nanos = measureNanoTime {
+            testRunCount = statisticsStorageLock.write {
+                updateStatisticsInternal(execution)
+            }
+        }
+
+        logger.info {
+            @Suppress(
+                "MagicNumber",
+                "FLOAT_IN_ACCURATE_CALCULATIONS",
+            )
+            "Test statistics updated (+$testRunCount test run(s)) in ${nanos / 1000L / 1e3} ms from execution (id = ${execution.requiredId()})"
+        }
+    }
+
+    /**
+     * Updates the statistics with test executions from this [execution].
+     *
+     * May be invoked concurrently from multiple threads, so should be
+     * externally guarded.
+     *
+     * @param execution either a historical or a newly-finished `Execution`.
+     * @return the number of test executions for this [execution].
+     * @see updateStatistics
+     */
+    @GuardedBy("statisticsStorageLock")
+    private fun updateStatisticsInternal(execution: Execution): Long {
+        val metadata = execution.metadata()
+
+        /*
+         * Process test executions in parallel (slightly faster).
+         */
+        return testExecutionRepository.findByExecutionId(execution.requiredId())
+            .parallelStreamIfEnabled(enabled = parallel)
+            .map { testExecution ->
+                val testId = metadata.extendWith(testExecution).let(testIdGenerator::testId)
+
+                statisticsStorage.updateExecutionStatistics(testId, testExecution.asTestRun())
+            }
+            .countEagerly()
+    }
+
+    /**
+     * May be invoked concurrently from multiple threads, so should be
+     * externally guarded.
+     *
      * @return the number of test executions for this [project].
+     * @see updateStatisticsInternal
      */
+    @GuardedBy("statisticsStorageLock")
     private fun updateStatistics(project: Project): Long {
         /*
          * Process executions sequentially.
          */
-        return executionService.getExecutionByNameAndOrganization(project.name, project.organization)
+        return executionRepository.getAllByProjectNameAndProjectOrganization(project.name, project.organization)
             .stream()
             .sequential()
             .sorted(comparingLong(Execution::requiredId))
-            .mapToLong { execution ->
-                /*
-                 * Process test executions in parallel (slightly faster).
-                 */
-                val testExecutions = testExecutionService.getTestExecutions(execution.requiredId())
-                    .parallelStream()
-
-                val metadataStream = execution.metadata().let { metadata ->
-                    Stream.generate { metadata }
```

Comment:
don't quite understand, seems, that previously we needed some few more actions
does `updateStatisticsInternal` cover all of them now?

Reply:
It's okay, don't worry.

I just simplified the code -- the metadata is now moved out of the stream pipeline and extracted into a separate variable. So zipping two streams is no longer necessary.

See [`updateStatisticsInternal()`](https://github.com/saveourtool/save-cloud/pull/1734/files#diff-e4c4ff9b5144fdc716635af988bdc8402baf9b84fc3c49447b47876f4597b198R232)